proposals/ssss-distributed-actor-runtime.md

```diff
@@ -0,0 +1,1798 @@
+    mutating func doneRecording() throws {
+      // our impl does not need to do anything here
+    }
+  }
+}
+```
+
+The above encoder is going to be called by the Swift runtime as was explained in the previous section.
+
+Once that is complete, the runtime will pass the constructed `InvocationEncoder` to the `remoteCall`:
+
+```swift
+ extension ClusterSystem {
+  // remoteCall is not a protocol requirement, however its signature is well known to the compiler,
+  // and it will invoke the method. We also are guaranteed that the 'Res: Codable' requirement is correct,
+  // since the type-system will enforce this conformance thanks to the type-level checks on distributed funcs.
+  func remoteCall<Act, Err, Res>(
+      on actor: Act,
+      target: RemoteCallTarget,
+      invocation: Self.InvocationEncoder,
+      throwing: Err.Type,
+      returning: Res.Type
+  ) async throws -> Res
+      where Act: DistributedActor,
+            Act.ID == ActorID.
+            Err: Error,
+            Res: Self.SerializationRequirement {
+    var envelope = invocation.envelope
+    
+    // [1] the recipient is transferred over the wire as its id
+    envelope.recipient = recipient.id
+      
+    // [2] the method is a mangled identifier of the 'distributed func' (or var).
+    //     In this system, we just use the mangled name, but we could do much better in the future.
+    envelope.target = target.mangledName
+      
+    // [3] send the envelope over the wire and await the reply:
+    let responseData = try await self.underlyingTransport.send(envelope, to: actor.id)
+
+    // [4] decode the response from the response bytes
+    // in our example system, we're using Codable as SerializationRequirement, 
+    // so we can decode the response like this (and never need to cast `as? Codable` etc.):
+    try self.someDecoder.decode(as: Res.self, from: data)
+  }
+}
+```
+
+The overall purpose of this `remoteCall` implementation is to create some form of message representation of the invocation and send it off to the remote node (or process) to receive and invoke the target method on the remote actor. 
+
+In our example implementation, the `Invocation` already serialized the arguments and stored them in the `Envelope`, so the `remoteCall` only needs to add the information about the call recipient `[1]`, and the target (method or computed property) of the call `[2]`. In our example implementation, we just store the target's mangled name `[2]`, which is simple, but it has its challenges in regard to protocol evolution.
+
+One notable issue that mangled names have is that any change in the method signature will result in not being able to resolve the target method anymore, we are very much aware of the issues this may cause to protocol evolution, and we lay out plans in [Future Directions](#future-directions) to improve the lookup mechanisms in ways that will even allow adding parameters (with default values), in wire (and ABI) compatible ways.
```

Comment:
```suggestion
One notable issue that mangled names have is that any change in the method signature will result in not being able to resolve the target method anymore. We are very much aware of the issues this may cause to protocol evolution, and we lay out plans in [Future Work](#future-work) to improve the lookup mechanisms in ways that will even allow adding parameters (with default values), in wire (and ABI) compatible ways.
```